include/watchdog.h

```diff
@@ -55,62 +70,237 @@ struct wdt_config {
+	u32_t min;
+	u32_t max;
+};
+
+/** Watchdog callback. */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
+
+/**
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window Timing parameters of watchdog timeout.
+ * @param handle Unique handle used to feed specific watchdog timeout.
+ *		 It is allowed to use the same handle for multistage timeouts.
+ *		 This handle is filled by the user, it will be supplied to
+ *		 callback. Good example is handle pointing to the id of the
+ *		 thread that is resposible for feeding corresponding timeout.
+ * @param callback Timeout callback. Passing NULL means that no callback
+ *		   will be run.
+ * @param next Pointer to the next timeout configuration. It is used for
+ *	       watchdogs with staged timeouts functionality. Should be NULL
+ *	       for single stage timeout.
+ * @param flags Bit field with following parts:
+ *
+ *     reset        [ 0 : 1 ]   - perform specified reset after timeout/callback
+ */
+struct wdt_timeout_cfg {
+	wdt_window window;
+	void *handle;
+	wdt_callback_t callback;
+#ifdef CONFIG_WDT_HAS_MULTISTAGE
+	struct wdt_timeout_cfg *next;
+#endif
+	u8_t flags;
+};
+
+/** @cond INTERNAL_HIDDEN */
 struct wdt_driver_api {
-	wdt_api_enable enable;
-	wdt_api_disable disable;
-	wdt_api_get_config get_config;
-	wdt_api_set_config set_config;
-	wdt_api_reload reload;
+	int (*setup)(struct device *dev, u8_t options);
+	int (*disable)(struct device *dev);
+	int (*install_timeout)(struct device *dev, struct wdt_timeout_cfg *cfg);
+	int (*feed)(struct device *dev, void *handle);
+	/** Deprecated functions */
+	void (*enable)(struct device *dev);
+	void (*disable)(struct device *dev);
+	void (*get_config)(struct device *dev, struct wdt_config *config);
+	int (*set_config)(struct device *dev, struct wdt_config *config);
+	void (*reload)(struct device *dev);
 };
+/**
+ * @endcond
+ */
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Set up watchdog instance.
+ *
+ * This function is used for configuring global watchdog settings that
+ * affect all timeouts. It should be called after installing timeouts.
+ * After successful return all installed timeouts are valid.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param options Bit field with following parts:
+ *
+ *     pause in sleep      [ 0 ]       - pause in CPU sleep
+ *     pause by debugger   [ 1 ]       - pause when halted by debugger
+ *
+ * @retval 0 If success.
+ * @retval -ENOTSUP If any of the set options is not supported.
+ * @retval -EBUSY If watchdog instance has been already setup.
+ */
+static inline int wdt_setup(struct device *dev, u8_t options)
+{
+	const struct wdt_driver_api *api = dev->driver_api;
+
+	return api->setup(dev, cfg);
+}
+
+/**
+ * @brief Disable watchdog instance.
+ *
+ * This function stops watchdog instance and automatically uninstalls all
+ * timeouts. To set up watchdog again it is required to install timeouts and
```

Comment:
To set up a new watchdog, install timeouts and call wdt_setup() again.
     Not all watchdogs can be restarted after they are disabled.

Note the use of the parens () to have doxygen create a link to the function name.